### PR TITLE
[Issue 63] Created Linking endpoint between user and project

### PR DESCRIPTION
**Pull request by thhu**
Added ability to associate user with project. This represents a many to many relationship. Also created APIs to look up users associated with a project and projects associated with an user.

issue link: https://github.com/SeattleVoluntech/seattlevoluntech/issues/63
pr: https://github.com/SeattleVoluntech/seattlevoluntech/pull/87

### CODE REVIEW

File: src/main/java/org/seattlevoluntech/controllers/UserController.java [hunk 1]
```diff
@@ -1,40 +1,42 @@
...
 public class UserController {
-
-    @Autowired UsersRepository userRepository;
     private final Logger logger = LoggerFactory.getLogger(this.getClass());
 
-    @RequestMapping("/users")
+    @Autowired
+    UsersRepository userRepository;
+
+    @GetMapping("/user")
```
Comment (FangJennyYuan) [inline src/main/java/org/seattlevoluntech/controllers/UserController.java]: should be "/users"

File: src/main/java/org/seattlevoluntech/controllers/UserController.java [hunk 2]
```diff
@@ -50,5 +52,18 @@ public String fetchDataByLastName(@RequestParam("last_name") String lastName){
         return result.toString();
     }
 
+    @GetMapping("/user/listProjects/{userId}")
```
Comment (FangJennyYuan) [inline src/main/java/org/seattlevoluntech/controllers/UserController.java]: Should be "/users/listProjects/{userId}"

File: src/main/java/org/seattlevoluntech/controllers/UserController.java (hunk 1 above)
Comment (Seiyaroo) [reply src/main/java/org/seattlevoluntech/controllers/UserController.java]: Looks accurate, I think this is pretty much what Tara and I were working towards but at a much slower pace haha. You guys are so fast!

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java [hunk 3]
```diff
@@ -86,6 +88,53 @@ public Project createProject(@RequestBody Project project){
...
+        {
+            throw new ErrorController.ErrorNotFound("userId-" + userId);
+        }
+
+
+        optionalProject.get().addUsers(optionalUser.get());
+        optionalUser.get().addProject(optionalProject.get());
+
+        projectsRepository.save(optionalProject.get());
+        usersRepository.save(optionalUser.get());
```
Comment (sscchan) [inline src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: Do we need to make sure the updates to projects and users tables are atomic?

File: src/main/java/org/seattlevoluntech/controllers/UserController.java (hunk 1 above)
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/UserController.java]: done

File: src/main/java/org/seattlevoluntech/controllers/UserController.java (hunk 2 above)
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/UserController.java]: done

File: src/main/java/org/seattlevoluntech/controllers/CallbackController.java [hunk 4]
```diff
@@ -34,16 +38,20 @@ public CallbackController() {
...
 
     @RequestMapping(value = "/callback", method = RequestMethod.POST, consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
     protected void postCallback(final HttpServletRequest req, final HttpServletResponse res) throws ServletException, IOException {
+        logger.info("post");
         handle(req, res);
     }
 
     private void handle(HttpServletRequest req, HttpServletResponse res) throws IOException {
         try {
+            logger.debug("request: " + req.toString());
```
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/CallbackController.java]: Sorry I didn't mean to commit this. Removed logging.

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java [hunk 5]
```diff
@@ -86,6 +88,53 @@ public Project createProject(@RequestBody Project project){
...
+        if (!optionalUser.isPresent())
+        {
+            throw new ErrorController.ErrorNotFound("userId-" + userId);
+        }
+
+
+        optionalProject.get().addUsers(optionalUser.get());
+        optionalUser.get().addProject(optionalProject.get());
+
+        projectsRepository.save(optionalProject.get());
```
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: I was following a guide and they did this, but let me test and confirm if we actually need both.

File: src/main/java/org/seattlevoluntech/controllers/UserController.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/UserController.java]: done

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -9,4 +9,7 @@ spring.datasource.username=${DB_USER}
 spring.datasource.password=${DB_PASS}
 
 #Do nothing
-spring.jpa.hibernate.ddl-auto=none
\ No newline at end of file
+#spring.jpa.hibernate.ddl-auto=none
+
+#Create tables in datasource.url if they do not exist
+spring.jpa.hibernate.ddl-auto=update
```
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/resources/application.properties]: whoops, was not suppose to commit this. Good catch.

File: src/main/java/org/seattlevoluntech/storage/User.java [hunk 7]
```diff
@@ -54,6 +36,14 @@ public User(
 
   private String bio;
 
+  @ManyToMany
+  @JoinTable(
+          name = "user_project",
```
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/storage/User.java]: done

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java (hunk 3 above)
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: That's a good point. I'm going to assume that we read more than write and use optimistic locking here. Let me know if you have any objections.

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: Tested, it works if you save the object on the entity that owns the relationship, in this case it's the users table.

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java (hunk 3 above)
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: Added optimistic locking on both user and project tables. However when manually inserting test data you have to ensure that the opt_lock field is filled with a integer.

File: src/main/java/org/seattlevoluntech/controllers/ProjectController.java [hunk 8]
```diff
@@ -86,6 +88,53 @@ public Project createProject(@RequestBody Project project){
...
+        usersRepository.save(optionalUser.get());
+
+        return projectsRepository.findById(projectId);
+    }
+
+    // Update project
+    @GetMapping(path="/projects/listUsers/{projectId}")
```
[the remark this answers is not in the document]
Comment (thhu) [reply src/main/java/org/seattlevoluntech/controllers/ProjectController.java]: Had an infinite loop issue where the user's object would load the projects which loads the users again. Fixed with @JsonIgnoreProperties.

**Pull request closed without merging**